tool/build/renamestr.c

```diff
@@ -0,0 +1,469 @@
+      Die(epath, "ELF PT_LOAD p_vaddr + p_memsz overflow");
+    }
+    if (p[i].p_offset + p[i].p_filesz < p[i].p_offset ||
+        p[i].p_offset + p[i].p_filesz + (pagesz - 1) < p[i].p_offset) {
+      Die(epath, "ELF PT_LOAD p_offset + p_filesz overflow");
+    }
+    if (p[i].p_align > 1 && ((p[i].p_vaddr & (p[i].p_align - 1)) !=
+                             (p[i].p_offset & (p[i].p_align - 1)))) {
+      Die(epath, "ELF p_vaddr incongruent w/ p_offset modulo p_align");
+    }
+    if ((p[i].p_vaddr & (pagesz - 1)) != (p[i].p_offset & (pagesz - 1))) {
+      Die(epath, "ELF p_vaddr incongruent w/ p_offset modulo AT_PAGESZ");
+    }
+    if (p[i].p_offset + p[i].p_filesz > esize) {
+      Die(epath, "ELF PT_LOAD p_offset and p_filesz overlaps image EOF");
+    }
+    Elf64_Off a = p[i].p_vaddr & -pagesz;
+    Elf64_Off b = (p[i].p_vaddr + p[i].p_memsz + (pagesz - 1)) & -pagesz;
+    for (j = i + 1; j < e->e_phnum; ++j) {
+      if (p[j].p_type != PT_LOAD)
+        continue;
+      Elf64_Off c = p[j].p_vaddr & -pagesz;
+      Elf64_Off d = (p[j].p_vaddr + p[j].p_memsz + (pagesz - 1)) & -pagesz;
+      if (MAX(a, c) < MIN(b, d)) {
+        Die(epath,
+            "ELF PT_LOAD phdrs %d and %d overlap each others virtual memory");
+      }
+    }
+    if (e->e_machine == EM_AARCH64 &&
+        p[i].p_vaddr + p[i].p_memsz > 0x0001000000000000) {
+      Die(epath, "this is an ARM ELF program but it loads to virtual"
+                 " memory addresses outside the legal range [0,2^48)");
+    }
+    if (e->e_machine == EM_NEXGEN32E &&
+        !(-140737488355328 <= (Elf64_Sxword)p[i].p_vaddr &&
+          (Elf64_Sxword)(p[i].p_vaddr + p[i].p_memsz) <= 140737488355328)) {
+      Die(epath, "this is an x86-64 ELF program, but it loads to virtual"
+                 " memory addresses outside the legal range [-2^47,2^47)");
+    }
+    if ((p[i].p_flags & PF_X) &&  //
+        p[i].p_vaddr <= e->e_entry &&
+        e->e_entry < p[i].p_vaddr + p[i].p_memsz) {
+      found_entry = 1;
+    }
+  }
+  if (!found_entry) {
+    Die(epath, "ELF entrypoint not found in PT_LOAD with PF_X");
+  }
+}
+
+struct Input {
+  union {
+    char *map;
+    Elf64_Ehdr *elf;
+    unsigned char *umap;
+  };
+  size_t size;
+  const char *path;
+};
+
+static struct Input input;
+
+static void OpenInput(const char *path) {
+  int fd;
+  if ((fd = open(path, O_RDWR)) == -1)
+    DieSys(path);
+  if ((input.size = lseek(fd, 0, SEEK_END)) == -1)
+    DieSys(path);
+  input.path = path;
+  input.map = mmap(0, input.size, PROT_READ | PROT_WRITE, MAP_PRIVATE, fd, 0);
+  if (input.map == MAP_FAILED)
+    DieSys(path);
+  if (!IsElf64Binary(input.elf, input.size))
+    Die(path, "not an elf64 binary");
+  exefd = fd;
+}
+
+static void ReplaceString(struct Param *param) {
+  Elf64_Xword len = strnlen(param->roloc, roend - param->roloc);
+  char *targstr = Malloc(len + 1);
+  for (Elf64_Xword i = 0; i < len + 1; ++i)
+    targstr[i] = 0;
```

Comment:
This could write out of bounds if the last string in a section isn't nul-terminated.

Reply:
Can this still happen with the `memmove` usage?